File: src/main/java/oatt/Intermediaries.java
package oatt;

abstract class Intermediaries {
    private String _name;

    void set_name(String name){
        _name = name;
    }

    String get_name(){
        return _name;
    }

    public String displayIntermediary(){
        return "";
    }
}

class Broker extends Intermediaries{
    private float _commission;

    public Broker(String name, float commission){
        set_name(name);
        set_commission(commission);
    }

    void set_commission(float commission){
        _commission = commission;
    }

    float get_commission(){
        return _commission;
    }

    public String displayIntermediary(){
        return "Broker: " + get_name() + ", Commission: " + get_commission();
    }
}

class Bank extends Intermediaries{
    private float _interestRate;

    public Bank(String name, float interestRate){
        set_name(name);
        set_interestRate(interestRate);
    }

    void set_interestRate(float interestRate){
        _interestRate = interestRate;
    }

    float get_interestRate(){
        return _interestRate;
    }

    public String displayIntermediary(){
        return "Bank: " + get_name() + ", Interest Rate: " + get_interestRate();
    }
}

class MututalFundManager extends Intermediaries{
    private String _employeeNumber;
    private float _managementFee;

    public MututalFundManager(String name, String employeeNumber, float managementFee){
        set_name(name);
        set_employeeNumber(employeeNumber);
        set_managementFee(managementFee);
    }

    void set_employeeNumber(String employeeNumber){
        _employeeNumber = employeeNumber;
    }

    void set_managementFee(float managementFee){
        _managementFee = managementFee;
    }

    String get_employeeNumber(){
        return _employeeNumber;
    }

    float get_managementFee(){
        return _managementFee;
    }

    public String displayIntermediary(){
        return "Mututal Fund Manager: " + get_name() + ", Employee Number: " + get_employeeNumber() + ", Management Fee: " + get_managementFee();
    }
}

File: Task-1/C++ Version/Intermediaries.cpp
#include "Intermediaries.h"
#include "SaveHelper.h"

void Intermediaries::set_name(const std::string& name) {
    _name = name;
}

std::string Intermediaries::get_name() const {
    return _name;
}

Broker::Broker(const std::string& name, float commission) {
    set_name(name);
    set_commission(commission);
}

void Broker::set_commission(float commission) {
    _commission = commission;
}

float Broker::get_commission() const {
    return _commission;
}

std::string Broker::displayIntermediary() const {
    return "Broker: " + get_name() + ", Commission: " + std::to_string(get_commission());
}

void Broker::serialize(std::ofstream& ofs) const {
    writeString(ofs, _name);
    ofs.write(reinterpret_cast<const char*>(&_commission), sizeof(_commission));
}

void Broker::deserialize(std::ifstream& ifs) {
    readString(ifs, _name);
    ifs.read(reinterpret_cast<char*>(&_commission), sizeof(_commission));
}

Bank::Bank(const std::string& name, float interestRate) {
    set_name(name);
    set_interestRate(interestRate);
}

void Bank::set_interestRate(float interestRate) {
    _interestRate = interestRate;
}

float Bank::get_interestRate() const {
    return _interestRate;
}

std::string Bank::displayIntermediary() const {
    return "Bank: " + get_name() + ", Interest Rate: " + std::to_string(get_interestRate());
}

void Bank::serialize(std::ofstream& ofs) const {
    writeString(ofs, _name);
    ofs.write(reinterpret_cast<const char*>(&_interestRate), sizeof(_interestRate));
}

void Bank::deserialize(std::ifstream& ifs) {
    readString(ifs, _name);
    ifs.read(reinterpret_cast<char*>(&_interestRate), sizeof(_interestRate));
}

MutualFundManager::MutualFundManager(const std::string& name, const std::string& employeeNumber, float managementFee) {
    set_name(name);
    set_employeeNumber(employeeNumber);
    set_managementFee(managementFee);
}

void MutualFundManager::set_employeeNumber(const std::string& employeeNumber) {
    _employeeNumber = employeeNumber;
}

void MutualFundManager::set_managementFee(float managementFee) {
    _managementFee = managementFee;
}

std::string MutualFundManager::get_employeeNumber() const {
    return _employeeNumber;
}

float MutualFundManager::get_managementFee() const {
    return _managementFee;
}

std::string MutualFundManager::displayIntermediary() const {
    return "Mutual Fund Manager: " + get_name() + ", Employee Number: " + get_employeeNumber() + ", Management Fee: " + std::to_string(get_managementFee());
}

void MutualFundManager::serialize(std::ofstream& ofs) const {
    writeString(ofs, _name);
    writeString(ofs, _employeeNumber);
    ofs.write(reinterpret_cast<const char*>(&_managementFee), sizeof(_managementFee));
}

void MutualFundManager::deserialize(std::ifstream& ifs) {
    readString(ifs, _name);
    readString(ifs, _employeeNumber);
    ifs.read(reinterpret_cast<char*>(&_managementFee), sizeof(_managementFee));
}